### PR TITLE
Add additional loss functions

### PR DESCRIPTION
**Pull request by AnzeXie**
- Added additional loss functions
- Put loss functions into separate files
- Created a new configuration section for loss functions
- Added `loss.reduction` option for loss functions
- Added corresponding documentation

### CODE REVIEW

File: src/cpp/include/loss.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
+#ifndef MARIUS_LOSS_H
+#define MARIUS_LOSS_H
+
+#include <torch/torch.h>
```
Comment (JasonMoho) [inline src/cpp/include/loss.h:4]: Follow the google style guide for header imports: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: src/cpp/include/decoder.h [hunk 2]
```diff
@@ -7,6 +7,7 @@
 
 #include "batch.h"
 #include "datatypes.h"
+#include "loss.h"
```
Comment (JasonMoho) [inline src/cpp/include/decoder.h:10]: Can this include statement be moved to `decoder.cpp`? It's best to keep includes out of the header files if they aren't needed in the header.

File: src/cpp/src/config.cpp [hunk 3]
```diff
@@ -277,14 +275,20 @@ MariusOptions parseConfig(int64_t argc, char *argv[]) {
...
     i_var_map.push_back((OptInfo<int>){&checkpoint_interval, 9999, "training", "checkpoint_interval", {1, INT32_MAX}});
     i_var_map.push_back((OptInfo<int>){&shuffle_interval, 1, "training", "shuffle_interval", {1, INT32_MAX}});
 
+    // LossOptions
+    LossFunctionType loss_function_type; string s_loss_function_type; // Loss to use
+    float margin; // Margin to use in ranking loss
+    ReductionType reduction_type; string s_reduction_type; // reduction method to use
+    s_var_map.push_back((OptInfo<std::string>){&s_loss_function_type, "SoftMax", "loss", "loss"});
+    f_var_map.push_back((OptInfo<float>){&margin, 0, "loss", "margin", {0, FLOAT_MAX}});
+    s_var_map.push_back((OptInfo<std::string>){&s_reduction_type, "Mean", "loss", "reduction"});
```
Comment (JasonMoho) [inline src/cpp/src/config.cpp:290]: The default should be Sum for SoftMax

File: src/cpp/python_bindings/loss_wrap.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#include <pybind11/pybind11.h>
```
Comment (JasonMoho) [inline src/cpp/python_bindings/loss_wrap.cpp:1]: Follow C++ style guide. https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: src/cpp/src/decoder.cpp [hunk 5]
```diff
@@ -224,30 +206,54 @@ void LinkPredictionDecoder::forward(Batch *batch, bool train) {
 }
 
 DistMult::DistMult() {
-    if (marius_options.training.loss_function_type == LossFunctionType::SoftMax) {
-        loss_function_ = new SoftMax();
-    } else if (marius_options.training.loss_function_type == LossFunctionType::RankingLoss) {
-        loss_function_ = new RankingLoss(marius_options.training.margin);
+    if (marius_options.loss.loss_function_type == LossFunctionType::SoftMax) {
```
Comment (JasonMoho) [inline src/cpp/src/decoder.cpp:209]: Can we move this into some function: `LossFunction *getLossFunction(LossFunctionType loss_function_type)`

That way we can reuse that function for all the Decoders

File: src/cpp/src/loss.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
+#include <loss.h>
```
Comment (JasonMoho) [inline src/cpp/src/loss.cpp:1]: Style guide

**Review by JasonMoho — CHANGES_REQUESTED** [review]

File: examples/training/configs/fb15k_cpu.ini [hunk 7]
```diff
@@ -23,11 +23,13 @@ negatives=512
 degree_fraction=.5
 learning_rate=.1
 optimizer=Adagrad
-loss=SoftMax
 synchronous=false
 num_epochs=10
 shuffle_interval=1
 
+[loss]
+loss=SoftMax
```
[the remark this answers is not in the document]
Comment (AnzeXie) [reply examples/training/configs/fb15k_cpu.ini:31]: Will update the tool in another pr

File: src/cpp/include/decoder.h (hunk 2 above)
Comment (AnzeXie) [reply src/cpp/include/decoder.h:10]: It seems this is still needed in the header file when declearing `LossFunction *loss_function_;`

**Comment by JasonMoho** [comment]
What is the status on this?

File: examples/training/configs/fb15k_cpu.ini (hunk 7 above)
[the remark this answers is not in the document]
Comment (AnzeXie) [reply examples/training/configs/fb15k_cpu.ini:31]: Updated all configuration files. The tool for modifications to configuration files is added in PR #49

**Review by JasonMoho — APPROVED** [review]
